### ISSUE TITLE
Openset KIE error

### ISSUE THREAD

**Issue by JohnReid** [open-mmlab/mmocr#634]
Thanks for your nice project, I am just trying to get to grips with KIE. I managed to run the closeset KIE but when I tried the openset version I get an error. My script looks like:

```python
from mmocr.utils.ocr import MMOCR

# Load models into memory
config = 'configs/kie/sdmgr/sdmgr_novisual_60e_wildreceipt_openset.py'
ocr = MMOCR(det='PS_CTW', recog='SAR', kie='SDMGR', kie_config=config)

# Inference
results = ocr.readtext('demo/demo_kie.jpeg', output='output/demo_kie.jpeg',
                       print_result=True, imshow=False,
                       export='export',
                       details=True)
print(results)
```

I get a `KeyError` in the output below though. I haven't managed to work out what I am doing wrong from the documentation :(.

```python
load checkpoint from http path: https://download.openmmlab.com/mmocr/textdet/psenet/psenet_r50_fpnf_600e_ctw1500_20210401-216fed50.pth
/media/data/Dev/mmocr/mmocr/apis/inference.py:48: UserWarning: Class names are not saved in the checkpoint's meta data, use COCO classes by default.
  warnings.warn('Class names are not saved in the checkpoint\'s '
load checkpoint from http path: https://download.openmmlab.com/mmocr/textrecog/sar/sar_r31_parallel_decoder_academic-dba3a4a3.pth
/media/data/Dev/mmocr/mmocr/apis/inference.py:48: UserWarning: Class names are not saved in the checkpoint's meta data, use COCO classes by default.
  warnings.warn('Class names are not saved in the checkpoint\'s '
load checkpoint from http path: https://download.openmmlab.com/mmocr/kie/sdmgr/sdmgr_unet16_60e_wildreceipt_20210520-7489e6de.pth
The model and loaded state dict do not match exactly

size mismatch for bbox_head.node_cls.weight: copying a param with shape torch.Size([26, 256]) from checkpoint, the shape in current model is torch.Size([4, 256]).
size mismatch for bbox_head.node_cls.bias: copying a param with shape torch.Size([26]) from checkpoint, the shape in current model is torch.Size([4]).
/home/john/anaconda3/envs/open-mmlab/lib/python3.7/site-packages/mmdet/datasets/utils.py:69: UserWarning: "ImageToTensor" pipeline is replaced by "DefaultFormatBundle" for batch inference. It is recommended to manually replace it in the test data pipeline in your config file.
  'data pipeline in your config file.', UserWarning)
/media/data/Dev/mmocr/mmocr/datasets/kie_dataset.py:46: UserWarning: KIEDataset is only initialized as a downstream demo task of text detection and recognition without an annotation file.
  'without an annotation file.', UserWarning)
/media/data/Dev/mmocr/mmocr/datasets/pipelines/box_utils.py:46: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  indexes = (np.arange(N, dtype=np.int) + lefttop_idx) % N
Traceback (most recent call last):
  File "./demo-kie", line 13, in <module>
    details=True)
  File "/media/data/Dev/mmocr/mmocr/utils/ocr.py", line 426, in readtext
    self.detect_model, self.recog_model, kie_model=self.kie_model)
  File "/media/data/Dev/mmocr/mmocr/utils/ocr.py", line 601, in det_recog_kie_inference
    batch_mode=self.args.batch_mode)
  File "/media/data/Dev/mmocr/mmocr/apis/inference.py", line 144, in model_inference
    data = test_pipeline(data)
  File "/home/john/anaconda3/envs/open-mmlab/lib/python3.7/site-packages/mmdet/datasets/pipelines/compose.py", line 41, in __call__
    data = t(data)
  File "/media/data/Dev/mmocr/mmocr/datasets/pipelines/kie_transforms.py", line 21, in __call__
    w, h = results['img_info']['width'], results['img_info']['height']
KeyError: 'img_info'
```

Thanks for any help!

**Comment by gaotongxiao** [comment]
Thanks for catching that. Our demo object doesn't support OpenSet KIE yet, though it technically should be. We'll look into that ASAP. Before we get it fixed, you can run
```bash
python tools/kie_test_imgs.py configs/kie/sdmgr/sdmgr_novisual_60e_wildreceipt_openset.py  --out-dir openset_kie
```
to get a sense of OpenSet KIE.

**Comment by JohnReid** [comment]
Thanks for the quick reply. That script requires a checkpoint arg:
```bash
(open-mmlab) ➜  mmocr git:(main) ✗ python tools/kie_test_imgs.py configs/kie/sdmgr/sdmgr_novisual_60e_wildreceipt_openset.py  --out-dir openset_kie

usage: kie_test_imgs.py [-h] [--show] [--out-dir OUT_DIR]
                        [--local_rank LOCAL_RANK] [--device DEVICE]
                        config checkpoint
kie_test_imgs.py: error: the following arguments are required: checkpoint
```

**Comment by gaotongxiao** [comment]
You can download the checkpoint from https://mmocr.readthedocs.io/en/latest/kie_models.html#wildreceiptopenset

**Comment by JohnReid** [comment]
Thanks, I had not worked that out but probably should have.

To get things working I additionally had to add `--device 0` to the command line otherwise I had a CPU / GPU error. Also it seems the `font_color` argument to `draw_texts_by_pil` has been changed to `fill_color` and I had to change that on lines 770 and 797 in `mmocr/core/visualize.py`. Happy to send a PR :)

**Comment by JohnReid** [comment]
Results looking nice though :)

**Comment by JohnReid** [comment]
Just FYI: I'm experiencing the same `img_info` error when trying to serve the model through torchserve. I guess that will all be part of the same fix.

**Comment by gaotongxiao** [comment]
> Thanks, I had not worked that out but probably should have.
> 
> To get things working I additionally had to add `--device 0` to the command line otherwise I had a CPU / GPU error. Also it seems the `font_color` argument to `draw_texts_by_pil` has been changed to `fill_color` and I had to change that on lines 770 and 797 in `mmocr/core/visualize.py`. Happy to send a PR :)

I didn't have any CPU/GPU error without `--device 0`, could you share the error message so that we can locate the bug? And the argument name change will be fixed in https://github.com/open-mmlab/mmocr/pull/610, anyway thanks for pointing out that.

> Just FYI: I'm experiencing the same `img_info` error when trying to serve the model through torchserve. I guess that will all be part of the same fix.

That's right :)

**Comment by JohnReid** [comment]
> > To get things working I additionally had to add `--device 0` to the command line otherwise I had a CPU / GPU error. Also it seems the `font_color` argument to `draw_texts_by_pil` has been changed to `fill_color` and I had to change that on lines 770 and 797 in `mmocr/core/visualize.py`. Happy to send a PR :)
> 
> I didn't have any CPU/GPU error without `--device 0`, could you share the error message so that we can locate the bug?
> 
I get this error
```bash
(open-mmlab) ➜  mmocr git:(main) ✗ python tools/kie_test_imgs.py configs/kie/sdmgr/sdmgr_novisual_60e_wildreceipt_openset.py checkpoints/sdmgr_novisual_60e_wildreceipt_openset_20210917-d236b3ea.pth --out-dir openset_kie
load checkpoint from local path: checkpoints/sdmgr_novisual_60e_wildreceipt_openset_20210917-d236b3ea.pth
[                                                  ] 0/472, elapsed: 0s, ETA:Traceback (most recent call last):
  File "tools/kie_test_imgs.py", line 161, in <module>
    main()
  File "tools/kie_test_imgs.py", line 157, in main
    test(model, data_loader, args.show, args.out_dir)
  File "tools/kie_test_imgs.py", line 59, in test
    result = model(return_loss=False, rescale=True, **data)
  File "/home/john/anaconda3/envs/open-mmlab/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/john/anaconda3/envs/open-mmlab/lib/python3.7/site-packages/mmcv/parallel/data_parallel.py", line 50, in forward
    return super().forward(*inputs, **kwargs)
  File "/home/john/anaconda3/envs/open-mmlab/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 156, in forward
    "them on device: {}".format(self.src_device_obj, t.device))
RuntimeError: module must have its parameters and buffers on device cuda:0 (device_ids[0]) but found one of them on device: cpu
```
My setup looks like:
```
(open-mmlab) ➜  mmocr git:(main) ✗ nvidia-smi
Fri Dec  3 09:08:59 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 00000001:00:00.0 Off |                    0 |
| N/A   61C    P0    63W / 149W |   3457MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           On   | 00000002:00:00.0 Off |                    0 |
| N/A   47C    P0    72W / 149W |   3457MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1125      G   /usr/lib/xorg/Xorg                  3MiB |
|    0   N/A  N/A      2396      C   ...nvs/open-mmlab/bin/python      537MiB |
|    0   N/A  N/A      2409      C   ...nvs/open-mmlab/bin/python      537MiB |
|    0   N/A  N/A      2429      C   ...nvs/open-mmlab/bin/python      728MiB |
|    0   N/A  N/A      2443      C   ...nvs/open-mmlab/bin/python      537MiB |
|    0   N/A  N/A      2457      C   ...nvs/open-mmlab/bin/python      537MiB |
|    0   N/A  N/A      2484      C   ...nvs/open-mmlab/bin/python      568MiB |
|    1   N/A  N/A      1125      G   /usr/lib/xorg/Xorg                  3MiB |
|    1   N/A  N/A      2395      C   ...nvs/open-mmlab/bin/python      537MiB |
|    1   N/A  N/A      2408      C   ...nvs/open-mmlab/bin/python      537MiB |
|    1   N/A  N/A      2430      C   ...nvs/open-mmlab/bin/python      728MiB |
|    1   N/A  N/A      2444      C   ...nvs/open-mmlab/bin/python      537MiB |
|    1   N/A  N/A      2458      C   ...nvs/open-mmlab/bin/python      537MiB |
|    1   N/A  N/A      2483      C   ...nvs/open-mmlab/bin/python      568MiB |
+-----------------------------------------------------------------------------+
```

**Comment by gaotongxiao** [comment]
It has something to do with `MMDataPrallel`, which managed to run the model on all of your GPUs. We will build a patch for that. Thanks again for reporting it!

**Comment by gaotongxiao** [comment]
It have been fixed in #752
